### PR TITLE
Fix finding Rterm on Windows

### LINKED ISSUE #787: M-x R in 18.10.2 does not work on Windows

### PR DESCRIPTION
**Pull request by jabranham**
'inferior-ess-r-program' might not point to an R executable,
especially on MS Windows, so we check it in 'run-ess-r'. If it's not
executable, set it to (ess-find-newest-R) which should be able to find
an R executable on any OS.

Closes #787

### CODE REVIEW

File: lisp/ess-r-mode.el [hunk 1]
```diff
@@ -917,6 +921,7 @@ Examples: (ess-current-R-at-least '2.7.0)
 (defun ess-newest-r (rvers)
```
Comment (vspinu) [inline lisp/ess-r-mode.el]: The  semantics of names `ess-find-newest-R` and `ess-newest-r` is unfortunate. I thought that the latter is actually running the newest R.  Shall we make the latter internal and name it more specifically, like `ess-r--newest-r-path`? Or just inline it whenever needed; it's a 2 liner. Capital R in one and not in the other is confusing.

File: lisp/ess-r-mode.el [hunk 2]
```diff
@@ -476,6 +481,12 @@ will be prompted to enter arguments interactively."
    (format
     "\n(R): ess-dialect=%s, buf=%s, start-arg=%s\n current-prefix-arg=%s\n"
     ess-dialect (current-buffer) start-args current-prefix-arg))
+  (unless (executable-find inferior-ess-r-program)
```
Comment (vspinu) [inline lisp/ess-r-mode.el]: This setq side effect here is not good. Such an important variable should be set in one place and not interfered with.

File: lisp/ess-r-mode.el [hunk 3]
```diff
@@ -252,6 +252,11 @@ When t, loading a file into a namespaced will output information
 about which objects are exported and which stay hidden in the
 namespace.")
 
+(defvar ess-newest-R nil
```
Comment (vspinu) [inline lisp/ess-r-mode.el]: Shall we make this one internal and make it more explicit like `ess-r--newest-executable` or something?

File: lisp/ess-r-mode.el (hunk 2 above)
Comment (vspinu) [reply lisp/ess-r-mode.el]: I think this logic can be used whenever needed without setting `inefrior-ess-r-program`.

File: lisp/ess-r-mode.el (hunk 1 above)
Comment (vspinu) [reply lisp/ess-r-mode.el]: Could be, but it then becomes semantically similar to `inferior-ess-r-program` which is not desirable.

Both `ess-newest-r` and the inner `ess-find-newest-date` are used once and have bad names. I would suggest we make a local `cl-labels` inside `ess-find-newest-R-version` and remove both of the above. Or at least inline `ess-find-newest-date` into `ess-newest-r` and rename the latter into an internal with a more suggestive name.

File: lisp/ess-r-mode.el (hunk 2 above)
Comment (jabranham) [reply lisp/ess-r-mode.el]: Agreed in principal, but I'm not sure the best way to handle it. One option is to put something like this in the defcustom, which will increase the time it takes to load ESS.

> this logic can be used whenever needed without setting inefrior-ess-r-program

probably, but it sounds like more work than I'm willing to put in for an evil, proprietary operating system. :-)

**Pull request closed without merging**